yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1059,9 +1059,11 @@ private[spark] class Client(
         } catch {
           case e: ApplicationNotFoundException =>
             logError(s"Application $appId not found.")
+            cleanupStagingDir(appId)
             return (YarnApplicationState.KILLED, FinalApplicationStatus.KILLED)
           case NonFatal(e) =>
```

Comment:
One scenario related to this issue might be connection failure between Client and AM in yarn cluster mode, in such scenario it is not safe to remove staging dir, probably Spark application is still running and it is OK local client is failed early.